File: internal/core/privacy.go
package core

import (
	"path/filepath"
	"strings"
)

type PrivacyChecker struct {
	Rules []string
}

func NewPrivacyChecker(rules []string) *PrivacyChecker {
	return &PrivacyChecker{Rules: rules}
}

func (p *PrivacyChecker) ShouldSkip(file FileRecord) bool {
	if file.AISkip {
		return true
	}
	// 匹配文件名或完整路径
	if p.Match(file.Name, file.LocalPath) {
		return true
	}
	// #18: 匹配父目录（用于排除整个目录如 confidential/）
	return p.Match("", filepath.Dir(file.LocalPath))
}

func (p *PrivacyChecker) Match(name string, path string) bool {
	for _, pattern := range p.Rules {
		if matchPattern(pattern, name) || matchPattern(pattern, path) {
			return true
		}
	}
	return false
}

func matchPattern(pattern, s string) bool {
	lower := strings.ToLower(s)
	p := strings.ToLower(pattern)

	if strings.Contains(p, "*") || strings.Contains(p, "?") {
		return globMatch(p, lower)
	}
	return strings.Contains(lower, p)
}

// globMatch 支持 * (单级) 和 ** (多级) 通配符。
func globMatch(pattern, s string) bool {
	// #17: 处理 ** 多级通配符
	if strings.Contains(pattern, "**") {
		return globMatchDoubleStar(pattern, s)
	}

	px, sx := 0, 0
	starIdx, matchIdx := -1, 0

	for sx < len(s) {
		if px < len(pattern) && (pattern[px] == s[sx] || pattern[px] == '?') {
			px++
			sx++
		} else if px < len(pattern) && pattern[px] == '*' {
			starIdx = px
			matchIdx = sx
			px++
		} else if starIdx != -1 {
			px = starIdx + 1
			matchIdx++
			sx = matchIdx
		} else {
			return false
		}
	}

	for px < len(pattern) && pattern[px] == '*' {
		px++
	}
	return px == len(pattern)
}

// globMatchDoubleStar handles ** patterns by splitting on ** and matching segments.
func globMatchDoubleStar(pattern, s string) bool {
	// Simple approach: replace ** with a special marker and use recursive matching
	segments := strings.Split(pattern, "**")
	if len(segments) == 1 {
		return globMatch(pattern, s)
	}

	// Each ** can match zero or more path segments
	return matchDoubleStarSegments(segments, s, 0, 0)
}

func matchDoubleStarSegments(segments []string, s string, segIdx, strIdx int) bool {
	if segIdx >= len(segments) {
		return strIdx >= len(s)
	}

	seg := segments[segIdx]
	isLast := segIdx == len(segments)-1

	// Try matching segment at every position from strIdx onwards
	for i := strIdx; i <= len(s); i++ {
		if globMatch(seg, s[strIdx:i]) {
			if isLast {
				// Last segment must match the remainder exactly
				if i == len(s) {
					return true
				}
			} else {
				if matchDoubleStarSegments(segments, s, segIdx+1, i) {
					return true
				}
			}
		}
	}
	return false
}

File: src-tauri/src/core/privacy.rs
use crate::core::models::FileRecord;
use std::path::Path;

pub struct PrivacyChecker {
    pub rules: Vec<String>,
}

impl PrivacyChecker {
    pub fn new(rules: Vec<String>) -> Self {
        Self { rules }
    }

    pub fn should_skip(&self, file: &FileRecord) -> bool {
        if file.ai_skip {
            return true;
        }
        if self.match_patterns(&file.name, &file.local_path) {
            return true;
        }
        // Match parent directory
        if let Some(parent) = Path::new(&file.local_path).parent() {
            if self.match_patterns("", &parent.to_string_lossy()) {
                return true;
            }
        }
        false
    }

    fn match_patterns(&self, name: &str, path: &str) -> bool {
        for pattern in &self.rules {
            if match_pattern(pattern, name) || match_pattern(pattern, path) {
                return true;
            }
        }
        false
    }
}

fn match_pattern(pattern: &str, s: &str) -> bool {
    let lower = s.to_lowercase();
    let p = pattern.to_lowercase();

    if p.contains('*') || p.contains('?') {
        glob_match(&p, &lower)
    } else {
        lower.contains(&p)
    }
}

fn glob_match(pattern: &str, s: &str) -> bool {
    if pattern.contains("**") {
        return glob_match_double_star(pattern, s);
    }

    let mut px = 0;
    let mut sx = 0;
    let mut star_idx = None;
    let mut match_idx = 0;
    let p: Vec<char> = pattern.chars().collect();
    let st: Vec<char> = s.chars().collect();

    while sx < st.len() {
        if px < p.len() && (p[px] == st[sx] || p[px] == '?') {
            px += 1;
            sx += 1;
        } else if px < p.len() && p[px] == '*' {
            star_idx = Some(px);
            match_idx = sx;
            px += 1;
        } else if let Some(si) = star_idx {
            px = si + 1;
            match_idx += 1;
            sx = match_idx;
        } else {
            return false;
        }
    }

    while px < p.len() && p[px] == '*' {
        px += 1;
    }

    px == p.len()
}

fn glob_match_double_star(pattern: &str, s: &str) -> bool {
    let segments: Vec<&str> = pattern.split("**").collect();
    if segments.len() == 1 {
        return glob_match(pattern, s);
    }
    match_double_star_segments(&segments, s, 0, 0)
}

fn match_double_star_segments(segments: &[&str], s: &str, seg_idx: usize, str_idx: usize) -> bool {
    if seg_idx >= segments.len() {
        return str_idx >= s.len();
    }

    let seg = segments[seg_idx];
    let is_last = seg_idx == segments.len() - 1;

    for i in str_idx..=s.len() {
        if glob_match(seg, &s[str_idx..i]) {
            if is_last {
                if i == s.len() {
                    return true;
                }
            } else if match_double_star_segments(segments, s, seg_idx + 1, i) {
                return true;
            }
        }
    }
    false
}
